File: problems/Longest Increasing Subsequence/lengthOfLIS.cs
public class Solution {
    public int LengthOfLIS(int[] nums) {
        return getLisLenDpAndBs(nums);
    }

    private int getLisLenDpAndBs(int[] nums) {
        int[] dp = new int[nums.Length];
        int result = 0;

        foreach (int num in nums) {
            int i = Array.BinarySearch(dp, 0, result, num);

            if (0 > i) {
                i = -(1 + i);
            }

            dp[i] = num;

            if (result == i) {
                ++result;
            }
        }

        return result;
    }

    private int getLisLenBottomUp(int[] nums) {
        if (null == nums || 0 == nums.Length) {
            return 0;
        }

        int n = nums.Length;
        int[] dp = new int[n];
        int result = 1;

        Array.Fill(dp, 1);

        for (int i = 1; n > i; ++i) {
            for (int j = 0; i > j; ++j) {
                if (nums[i] > nums[j] && dp[i] <= dp[j]) {
                    dp[i] = 1 + dp[j];
                    result = Math.Max(result, dp[i]);
                }
            }
        }

        return result;
    }
}

File: problems/Longest Increasing Subsequence/lengthOfLIS.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        Integer[][] dp = new Integer[n][1 + n];

        return getLisTopDown(nums, dp, 0, -1);
    }

    private int getLisTopDown(int[] nums, Integer[][] dp, int currIdx, int prevIdx) {
        if (nums.length == currIdx) {
            return 0;
        }

        if (null == dp[currIdx][1 + prevIdx]) {
            int lenWith = -1 == prevIdx || nums[currIdx] > nums[prevIdx] ? 1 + getLisTopDown(nums, dp, 1 + currIdx, currIdx) : 0;
            int lenWithout = getLisTopDown(nums, dp, 1 + currIdx, prevIdx);

            dp[currIdx][1 + prevIdx] = Math.max(lenWith, lenWithout);
        }

        return dp[currIdx][1 + prevIdx];
    }

    private int getLisLenRecursive(int[] nums, int currIdx, int prevIdx) {
        if (nums.length == currIdx) {
            return 0;
        }

        int lenWith = -1 == prevIdx || nums[currIdx] > nums[prevIdx] ? 1 + getLisLenRecursive(nums, 1 + currIdx, currIdx) : 0;
        int lenWithout = getLisLenRecursive(nums, 1 + currIdx, prevIdx);

        return Math.max(lenWith, lenWithout);
    }
}
